### ISSUE TITLE
wgpu backend: powf_primitive redefined in shader during backward pass

### ISSUE THREAD

**Issue by takabo0222** [tracel-ai/burn#3329]
**Describe the bug**  
When performing a backward pass (`loss.backward()`) using the `wgpu` backend in Burn v0.17.1, a shader validation error occurs. The error message indicates that the function `powf_primitive` is being redefined in the generated WGSL shader. This leads to a runtime panic during shader compilation.

This issue does **not** occur in Burn v0.16.1, suggesting a regression introduced in version 0.17.1 related to WGSL shader code generation.

---

**To Reproduce**  

1. Create a new Rust library project:  
   ```bash
   cargo new burn_repro_issue --lib

Replace Cargo.toml with:

   ```Cargo.toml
   [package]
   name = "burn_repro_issue"
   version = "0.1.0"
   edition = "2021"

   [dependencies]
   burn = { version = "0.17.1", features = ["train", "wgpu"] }
   ```

Replace src/lib.rs with the provided minimal reproducible example (MRE) that defines:

An AlphaZero-style model

A custom loss function using .log()

A test case that triggers a backward pass

```rust
#![cfg(test)]

mod tests {
    use burn::{
        prelude::*,
        backend::{Wgpu, Autodiff},
        nn::{
            conv::{Conv2d, Conv2dConfig},
            loss::{CrossEntropyLoss, MseLoss, Reduction},
            BatchNorm, BatchNormConfig, Linear, LinearConfig, Relu, Tanh,
        },
        tensor::{self, activation::softmax, f16},
    };

    // --- Start: Copied from `types/constant.rs` ---
    pub const D: usize = 2;
    pub const H: usize = 8;
    pub const W: usize = 8;
    pub const BATCH_SIZE: usize = 4;
    pub const IN_CHANNEL: usize = 2;
    pub const RES_CHANNEL: usize = 16;
    pub const RES_BLOCKS: usize = 2;
    // --- End: Copied from `types/constant.rs` ---

    // --- Start: Copied from `types/agent.rs` ---
    #[derive(Clone)]
    pub struct Input<B: Backend, const C: usize> {
        pub data: Tensor<B, 4>,
        pub legal: Tensor<B, 3>,
    }

    impl<B: Backend, const C: usize> Input<B, C> {
        pub fn new(data: Tensor<B, 4>, legal: Tensor<B, 3>) -> Result<Self, String> {
            let [b, c, _h, _w] = data.dims();
            if c != C { return Err(format!("Channel mismatch: expected {}, got {}", C, c)); }
            let [lb, _, _] = legal.dims();
            if b != lb { return Err(format!("Batch size mismatch: data {}, legal {}", b, lb)); }
            Ok(Self { data, legal })
        }
    }

    #[derive(Clone, Debug)]
    pub struct ValueReturn<B: Backend> {
        pub data: Tensor<B, 2>,
    }

    impl<B: Backend> ValueReturn<B> {
        pub fn new(data: Tensor<B, 2>) -> Result<Self, String> {
            let [_, v] = data.dims();
            if v != 1 { return Err(format!("Value dimension mismatch: expected 1, got {}", v)); }
            Ok(Self { data })
        }
    }

    #[derive(Clone, Debug)]
    pub struct PolicyReturn<B: Backend> {
        pub data: Tensor<B, 2>,
    }

    impl<B: Backend> PolicyReturn<B> {
        pub fn new(data: Tensor<B, 2>) -> Result<Self, String> {
            let [_, c] = data.dims();
            if c != 64 { return Err(format!("Policy dimension mismatch: expected 64, got {}", c)); }
            Ok(Self { data })
        }
    }
    // --- End: Copied from `types/agent.rs` ---

    // --- Start: Copied from `agent/model/pretreat.rs` ---
    #[derive(Module, Debug)]
    pub struct Pretreat<B: Backend, const IC: usize, const OC: usize> {
        pub conv: Conv2d<B>,
        batch_norm: BatchNorm<B, D>,
        activation: Relu,
    }

    #[derive(Config, Debug)]
    pub struct PretreatConfig {
        pub kernel_size: [usize; 2],
    }

    impl PretreatConfig {
        pub fn init<B: Backend, const IC: usize, const OC: usize>(&self, device: &B::Device) -> Pretreat<B, IC, OC> {
            Pretreat {
                conv: Conv2dConfig::new([IC, OC], self.kernel_size).with_padding(nn::PaddingConfig2d::Same).init(device),
                batch_norm: BatchNormConfig::new(OC).init(device),
                activation: Relu::new(),
            }
        }
    }
    
    impl<B: Backend, const IC: usize, const OC: usize> Pretreat<B, IC, OC> {
        pub fn forward(&self, input: Input<B, IC>) -> Input<B, OC> {
            let x = self.conv.forward(input.data);
            let x = self.batch_norm.forward(x);
            let x = self.activation.forward(x);
            Input::new(x, input.legal).unwrap()
        }
    }
    // --- End: Copied from `agent/model/pretreat.rs` ---

    // --- Start: Copied from `agent/model/residual.rs` ---
    #[derive(Module, Debug)]
    pub struct Residual<B: Backend, const C: usize> {
        conv1: Conv2d<B>,
        batch_norm1: BatchNorm<B, D>,
        conv2: Conv2d<B>,
        batch_norm2: BatchNorm<B, D>,
        activation: Relu,
    }

    #[derive(Config, Debug)]
    pub struct ResidualConfig {
        pub kernel_size: [usize; 2],
    }

    impl ResidualConfig {
        pub fn init<B: Backend, const C: usize>(&self, device: &B::Device) -> Residual<B, C> {
            Residual {
                conv1: Conv2dConfig::new([C, C], self.kernel_size).with_padding(nn::PaddingConfig2d::Same).init(device),
                batch_norm1: BatchNormConfig::new(C).init(device),
                conv2: Conv2dConfig::new([C, C], self.kernel_size).with_padding(nn::PaddingConfig2d::Same).init(device),
                batch_norm2: BatchNormConfig::new(C).init(device),
                activation: Relu::new(),
            }
        }
    }

    impl<B: Backend, const C: usize> Residual<B, C> {
        pub fn forward(&self, input: Input<B, C>) -> Input<B, C> {
            let x = self.conv1.forward(input.data.clone());
            let x = self.batch_norm1.forward(x);
            let x = self.conv2.forward(x);
            let x = self.batch_norm2.forward(x);
            let x = x + input.data;
            let x = self.activation.forward(x);
            Input::new(x, input.legal).unwrap()
        }
    }
    // --- End: Copied from `agent/model/residual.rs` ---

    // --- Start: Copied from `agent/model/value_net.rs` ---
    #[derive(Module, Debug)]
    pub struct ValueNet<B: Backend, const IC: usize> {
        conv1: Conv2d<B>,
        batch_norm: BatchNorm<B, D>,
        linear1: Linear<B>,
        linear2: Linear<B>,
        activation: Relu,
        activation_output: Tanh,
    }

    #[derive(Config, Debug)]
    pub struct ValueNetConfig {
        pub kernel_size: [usize; 2],
        pub hidden_size: usize,
    }

    impl ValueNetConfig {
        pub fn init<B: Backend, const IC: usize, const OC: usize>(&self, device: &B::Device) -> ValueNet<B, IC> {
            const OUTPUT: usize = 1;
            ValueNet {
                conv1: Conv2dConfig::new([IC, OC], self.kernel_size).init(device),
                batch_norm: BatchNormConfig::new(OC).init(device),
                linear1: LinearConfig::new(OC * H * W, self.hidden_size).init(device),
                linear2: LinearConfig::new(self.hidden_size, OUTPUT).init(device),
                activation: Relu::new(),
                activation_output: Tanh::new(),
            }
        }
    }

    impl<B: Backend, const IC: usize> ValueNet<B, IC> {
        pub fn forward(&self, input: Input<B, IC>) -> ValueReturn<B> {
            let x = self.conv1.forward(input.data);
            let [batch_size, channel, height, width] = x.dims();
            let x = self.batch_norm.forward(x);
            let x = self.activation.forward(x);
            let x = x.reshape([batch_size, channel * height * width]);
            let x = self.linear1.forward(x);
            let x = self.activation.forward(x);
            let x = self.linear2.forward(x);
            let x = self.activation_output.forward(x);
            ValueReturn::new(x).unwrap()
        }
    }
    // --- End: Copied from `agent/model/value_net.rs` ---

    // --- Start: Copied from `agent/model/policy_net.rs` ---
    #[derive(Module, Debug)]
    pub struct PolicyNet<B: Backend, const IC: usize> {
        conv: Conv2d<B>,
        batch_norm: BatchNorm<B, D>,
        linear: Linear<B>,
        activation: Relu,
    }

    #[derive(Config, Debug)]
    pub struct PolicyNetConfig {
        pub kernel_size: [usize; 2],
    }

    impl PolicyNetConfig {
        pub fn init<B: Backend, const IC: usize, const OC: usize>(&self, device: &B::Device) -> PolicyNet<B, IC> {
            const CLASSES: usize = 64;
            PolicyNet {
                conv: Conv2dConfig::new([IC, OC], self.kernel_size).init(device),
                batch_norm: BatchNormConfig::new(OC).init(device),
                linear: LinearConfig::new(OC * H * W, CLASSES).init(device),
                activation: Relu::new(),
            }
        }
    }

    impl<B: Backend, const IC: usize> PolicyNet<B, IC> {
        pub fn forward(&self, input: Input<B, IC>) -> PolicyReturn<B> {
            let x = self.conv.forward(input.data);
            let [batch_size, channel, height, width] = x.dims();
            let x = self.batch_norm.forward(x);
            let x = self.activation.forward(x);
            let x = x.reshape([batch_size, channel * height * width]);
            let x = self.linear.forward(x);
            let x = Self::masked_softmax(x, input.legal);
            PolicyReturn::new(x).unwrap()
        }

        fn masked_softmax(policy: Tensor<B, 2>, legal: Tensor<B, 3>) -> Tensor<B, 2> {
            let [batch_size, height, width] = legal.dims();
            let legal = legal.reshape([batch_size, height * width]);
            let device = &B::Device::default();
            let zeros = Tensor::zeros(legal.dims(), device);
            let mask = legal.equal(zeros);
            let neg_inf = Tensor::<B, 2>::from_floats([[f16::NEG_INFINITY]], device).expand(policy.dims());
            let masked_policy = policy.mask_where(mask, neg_inf);
            softmax(masked_policy, 1)
        }
    }
    // --- End: Copied from `agent/model/policy_net.rs` ---

    // --- Start: Copied from `agent/model/alphazero_model.rs` ---
    #[derive(Module, Debug)]
    pub struct AlphaZeroModel<B: Backend, const IC: usize, const RC: usize, const L: usize> {
        pretreat: Pretreat<B, IC, RC>,
        resnet: [Residual<B, RC>; L],
        value: ValueNet<B, RC>,
        policy: PolicyNet<B, RC>,
    }

    #[derive(Config, Debug)]
    pub struct AlphaZeroModelConfig {
        #[config(default = "[3, 3]")]
        pretreat_kernel_size: [usize; 2],
        #[config(default = "[3, 3]")]
        residual_kernel_size: [usize; 2],
        #[config(default = "[1, 1]")]
        value_kernel_size: [usize; 2],
        #[config(default = "256")]
        value_hidden_size: usize,
        #[config(default = "[1, 1]")]
        policy_kernel_size: [usize; 2],
    }

    impl AlphaZeroModelConfig {
        pub fn init<B: Backend, const IC: usize, const RC: usize, const L: usize>(&self, device: &B::Device) -> AlphaZeroModel<B, IC, RC, L> {
            let pretreat_config = PretreatConfig { kernel_size: self.pretreat_kernel_size };
            let residual_config = ResidualConfig { kernel_size: self.residual_kernel_size };
            let value_config = ValueNetConfig { kernel_size: self.value_kernel_size, hidden_size: self.value_hidden_size };
            let policy_config = PolicyNetConfig { kernel_size: self.policy_kernel_size };

            AlphaZeroModel {
                pretreat: pretreat_config.init(device),
                resnet: std::array::from_fn(|_| residual_config.init(device)),
                value: value_config.init::<B, RC, 1>(device),
                policy: policy_config.init::<B, RC, 2>(device),
            }
        }
    }

    impl<B: Backend, const IC: usize, const RC: usize, const L: usize> AlphaZeroModel<B, IC, RC, L> {
        pub fn forward(&self, input: Input<B, IC>) -> (ValueReturn<B>, PolicyReturn<B>) {
            let mut x = self.pretreat.forward(input);
            for block in &self.resnet {
                x = block.forward(x);
            }
            let value = self.value.forward(x.clone());
            let policy = self.policy.forward(x);
            (value, policy)
        }
    }

    #[derive(Module, Debug)]
    pub struct AlphaZeroLoss<B: Backend> {
        loss_v: MseLoss,
        loss_p: CrossEntropyLoss<B>,
    }

    impl<B: Backend> AlphaZeroLoss<B> {
        pub fn new(device: &B::Device) -> Self {
            Self {
                loss_v: MseLoss::new(),
                loss_p: CrossEntropyLoss::new(None, device),
            }
        }

        pub fn forward(&self, output_v: ValueReturn<B>, output_p: PolicyReturn<B>, target_v: ValueReturn<B>, target_p: PolicyReturn<B>) -> Tensor<B, 1> {
            let loss_v = self.loss_v.forward(output_v.data, target_v.data, Reduction::Mean);
            let loss_p = -(target_p.data * output_p.data.log()).sum_dim(1).mean();
            loss_p + loss_v
        }
    }
    // --- End: Copied from `agent/model/alphazero_model.rs` ---

    #[test]
    fn test_backward_pass_repro() {
        type B = Wgpu;
        type AD = Autodiff<B>;

        let device = Default::default();
        let config = AlphaZeroModelConfig::new();
        let model = config.init::<AD, IN_CHANNEL, RES_CHANNEL, RES_BLOCKS>(&device);

        let input_data = Tensor::<AD, 4>::zeros([BATCH_SIZE, IN_CHANNEL, H, W], &device);
        let legal_moves = Tensor::<AD, 3>::ones([BATCH_SIZE, H, W], &device);
        let input = Input::new(input_data, legal_moves).unwrap();

        let (output_v, output_p) = model.forward(input);

        let target_v_data = Tensor::<AD, 2>::zeros([BATCH_SIZE, 1], &device);
        let target_p_data = Tensor::<AD, 2>::zeros([BATCH_SIZE, H * W], &device);
        let target_v = ValueReturn::new(target_v_data).unwrap();
        let target_p = PolicyReturn::new(target_p_data).unwrap();

        let loss_fn = AlphaZeroLoss::<AD>::new(&device);
        let loss = loss_fn.forward(output_v, output_p, target_v, target_p);

        let grads = loss.backward();

        let pretreat_grad = model.pretreat.conv.weight.grad(&grads);
        assert!(pretreat_grad.is_some());
        println!("Successfully reproduced backward pass.");
    }
}
```

Run the test:
   ```bash
   cargo test
   ```

The test fails with the following error:
```
wgpu error: Validation Error
Shader '' parsing error: redefinition of `powf_primitive`
┌─ wgsl:142:4
│
142 │ fn powf_primitive(lhs: f32, rhs: f32) -> f32 {
    │    ^^^^^^^^^^^^^^ previous definition
...
167 │ fn powf_primitive(lhs: f32, rhs: f32) -> f32 {
    │    ^^^^^^^^^^^^^^ redefinition
```
Expected behavior
The backward pass should complete without any shader compilation errors. The model should be differentiable as expected using the wgpu backend.

Screenshots
N/A

Desktop:

OS: Windows

Burn version: 0.17.1

Backend: wgpu (Autodiff<Wgpu>)

Smartphone:
N/A

Additional context

This issue is triggered specifically by a model that includes multiple Conv2d, BatchNorm, and residual layers.

The .log() operation in the loss function appears to contribute to the shader function duplication.

Downgrading Burn to version 0.16.0 resolves the issue, suggesting a regression in shader function deduplication or code generation.

**Comment by nathanielsimard** [comment]
I was able to reproduce it on burn 0.17.1, but it seems fixed on main.

**Comment by takabo0222** [comment]
Thank you for testing.
I was able to confirm that it worked properly with hash #232f1b16.
